pkg/mutation/mutation_function_test.go

```diff
@@ -0,0 +1,354 @@
+					if portAsMap["name"] == "portName2B" {
+						if portAsMap["hostIP"] != TestValue {
+							t.Errorf("Failed to update pod")
+						}
+					} else {
+						if _, ok := port.(map[string]interface{})["hostIP"]; ok {
+							t.Errorf("Unexpected pod was updated")
+						}
+					}
+				}
+			} else {
+				for _, port := range container.(map[string]interface{})["ports"].([]interface{}) {
+					if _, ok := port.(map[string]interface{})["hostIP"]; ok {
+						t.Errorf("Unexpected pod was updated")
+					}
+				}
+			}
+		}
+	}
+
+	if err := testAssignMutation(
+		`spec.containers["name": "testname2"].ports["name": "portName2B"].hostIP`,
+		ParameterTestValue,
+		prepareTestPod(t),
+		testFunc,
+		t,
+	); err != nil {
+		t.Errorf("Unexpected error: %+v", err)
+	}
+}
+
+func TestListsAsLastElementWithStringValue(t *testing.T) {
+	testFunc := func(u *unstructured.Unstructured) {}
+
+	if err := testAssignMutation(
+		`spec.containers["name": "notExists"]`,
+		ParameterTestValue,
+		prepareTestPod(t),
+		testFunc,
+		t,
+	); err == nil {
+		t.Errorf("List path entry in last position should accept a string value")
+	}
+}
+
+func TestListsAsLastElement(t *testing.T) {
+	testFunc := func(u *unstructured.Unstructured) {
+		containers, _, err := unstructured.NestedFieldNoCopy(u.Object, "spec", "containers")
+		if err != nil {
+			t.Error("Unexpected error", err)
+		}
+		for _, container := range containers.([]interface{}) {
+			if container.(map[string]interface{})["name"] == "notExists" {
+				if container.(map[string]interface{})["foo"] == "bar" {
+					return
+				}
+			}
+		}
+		t.Errorf("Failed to update pod. Container element was not added")
+	}
+
+	if err := testAssignMutation(
+		`spec.containers["name": "notExists"]`,
+		"{\"name\": \"notExists\", \"foo\": \"bar\"}",
+		prepareTestPod(t),
+		testFunc,
+		t,
+	); err != nil {
+		t.Errorf("Unexpected error: %+v", err)
+	}
+}
+
+func TestListsAsLastElementAlreadyExists(t *testing.T) {
+	testFunc := func(u *unstructured.Unstructured) {
+		containers, _, err := unstructured.NestedSlice(u.Object, "spec", "containers")
+		if err != nil {
+			t.Error("Unexpected error", err)
+		}
+		for _, container := range containers {
+			if container.(map[string]interface{})["name"] == "testname1" {
+				if container.(map[string]interface{})["foo"] == "bar" {
+					return
+				}
+			}
+		}
+		t.Errorf("Pod's container was not replaced as expected")
+	}
+
+	if err := testAssignMutation(
+		`spec.containers["name": "testname1"]`,
+		"{\"name\": \"testname1\", \"foo\": \"bar\"}",
+		prepareTestPod(t),
+		testFunc,
+		t,
+	); err != nil {
+		t.Errorf("Unexpected error: %+v", err)
+	}
+}
+
+func TestListsAsLastElementAlreadyExistsWithKeyConflict(t *testing.T) {
+	testFunc := func(u *unstructured.Unstructured) {}
+	if err := testAssignMutation(
+		`spec.containers["name": "testname1"]`,
+		"{\"name\": \"conflictingName\", \"foo\": \"bar\"}",
+		prepareTestPod(t),
+		testFunc,
+		t,
+	); err == nil {
```

Comment:
Shouldn't this test be looking for an error? I don't see where we are expecting an error. We should also make sure that the error we're getting is the one we're expecting (i.e. caused by a failed validation and not a bug)

Reply:
Note that this intesects with the tests for conflicting keys on mutation validation that exist in another PR. I could see that making merging this difficult.

Maybe there can be some internal function to bypass validation?

Reply:
> Shouldn't this test be looking for an error? 
It actually does. It fail when the error is not there (err==nil), but the fail message was wrong. Updated the message, also looking for the specific error.

> Note that this intesects with the tests 
We could add an override of the path in AssignMutator:
`func (m AssignMutator) SetPath([]PathEntry)`
to update the created mutator before the mutation function.

Reply:
That SGTM. Should we make it a private function?

Reply:
Mutators are in a different package, so the function would have to be public.
I added a custom Mutator in the test (to avoid a public function for tests only), and added a test using it: TestListsAsLastElementAlreadyExistsWithKeyConflict_WithCustomMutator 
I kept the original test for now, with a comment to remove it after mutation validation is added (imho better than deleting it right away, but I don't mind getting rid of it immediately either).

This is all quite an exercise just to avoid a public function, so let me know if you think that would be the nicer solution.